File: src/main/java/com/namu/skill/SkillPlugin.java
package com.namu.skill;

import com.namu.skill.model.Dash;
import com.namu.skill.model.Skill;
import com.namu.skill.model.ThrowSword;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Arrays;
import java.util.HashMap;

public final class SkillPlugin extends JavaPlugin implements CommandExecutor {

    private static SkillPlugin instance;

    public static SkillPlugin getInstance() {
        return instance;
    }

    @Override
    public void onEnable() {
        instance = this;

        // register Skill
        skillHashMap.put("Dash", new Dash());
        skillHashMap.put("ThrowSword", new ThrowSword());

        // register command
        getCommand("skill").setExecutor(this);
    }

    @Override
    public void onDisable() {
    }

    private HashMap<String, Skill> skillHashMap = new HashMap<>();

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (args.length == 0) {
            sender.sendMessage(Arrays.toString(skillHashMap.keySet().toArray(new String[0])));
            return false;
        }

        Skill skill = skillHashMap.get(args[0]);
        Skill.power = Double.parseDouble(args[1]);
        Skill.x = Double.parseDouble(args[2]);
        Skill.y = Double.parseDouble(args[3]);
        Skill.z = Double.parseDouble(args[4]);

        if (skill == null) {
            sender.sendMessage("Unknown Skill!");
            return false;
        }

        skill.use((Player) sender);
        return true;
    }
}

File: src/main/java/com/namu/skill/SkillPlugin.kt
package com.namu.skill

import com.kkomi.devlibrary.command.CommandManager
import com.kkomi.devlibrary.nms.getNBTTagCompound
import com.namu.skill.command.CommandDebug
import com.namu.skill.command.CommandGetSkillIcon
import com.namu.skill.command.CommandUseSkill
import com.namu.skill.model.SkillManager
import com.namu.skill.model.SkillNameTag
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.plugin.java.JavaPlugin

class SkillPlugin : JavaPlugin(), CommandExecutor, Listener {

    companion object {
        lateinit var instance: SkillPlugin
    }

    override fun onEnable() {
        instance = this

        CommandManager(true).apply {
            addComponent("icon", CommandGetSkillIcon())
            addComponent("use", CommandUseSkill())
            addComponent("debug", CommandDebug())
        }.register(getCommand("skill"))

        Bukkit.getPluginManager().registerEvents(this, this)
    }

    override fun onDisable() {}

    @EventHandler
    fun onPlayerInteractEvent(event: PlayerInteractEvent) {
        if (event.action == Action.RIGHT_CLICK_AIR || event.action == Action.RIGHT_CLICK_BLOCK) {
            val itemStack = event.item ?: return
            val skillNameTag = itemStack.getNBTTagCompound(SkillNameTag::class.java) ?: return

            SkillManager.getSkill(skillNameTag.name)?.use(event.player)
        }
    }

}